### ISSUE TITLE
[inductor] `_precompile_worker` should catch Triton `CompilationError` and skip invalid autotuning configs

### ISSUE THREAD

**Issue by XAheli** [pytorch/pytorch#190002]
### Description

When Inductor generates autotuning configs for persistent-reduction or combo kernels, some configs can produce Triton code with shape mismatches (e.g., mask `[XBLOCK, 1]` vs pointer `[1, RBLOCK]`). During the original `torch.compile` run, the autotuner selects a working config and everything is fine. But during **cold import** of AOT-serialized code (or when Triton's warm cache is missing), `_precompile_worker` tries to compile **all** configs ahead of time — and crashes on the first invalid one.

The root cause is that `_precompile_worker` (in `torch/_inductor/runtime/triton_heuristics.py`, line ~762) only catches `(OutOfResources, PTXASError, IntelGPUError)`. A Triton `CompilationError` — which wraps things like `ValueError` from broadcast shape mismatches, "Loop-carried variable" errors, or other Triton IR issues — propagates uncaught and kills
the entire import.

Similarly, `_make_launcher` (line ~1005) only catches `(OutOfResources, PTXASError, torch.cuda.OutOfMemoryError, IntelGPUError)`.

This is inconsistent with `scheduler.py`, which already imports and catches `CompilationError` from `triton.compiler.errors` in two places (lines ~5970 and ~9931) as a workaround for triton-lang/triton#2151.

### Steps to reproduce

The simplest way to trigger this is to have a kernel where `tl.atomic_add` is called with a pointer that only depends on the reduction index (shape `[1, RBLOCK]`) but the mask depends on the x-dimension (shape `[XBLOCK, 1]`). When `XBLOCK > 1`, Triton's `atom_red_typechecking_impl` uses `broadcast_impl_shape` (one-way broadcast) which fails.

Even without the atomic_add case, any Triton `CompilationError` during config precompilation crashes the import instead of gracefully skipping that config.

**Minimal reproducer showing the exception handling gap:**

```python
"""
Demonstrates that _precompile_worker does not catch CompilationError.

This script manually calls CachingAutotuner._precompile_worker() with a kernel that has one valid config and one invalid config, showing that the invalid config's CompilationError crashes everything.
"""
import torch
from triton.compiler.errors import CompilationError

# The exception types caught by _precompile_worker
from triton.runtime.errors import OutOfResources

print("OutOfResources would be caught:", issubclass(OutOfResources, (OutOfResources,)))
print("CompilationError would be caught:", issubclass(CompilationError, (OutOfResources,)))
print()
print("CompilationError MRO:", [c.__name__ for c in CompilationError.__mro__])
print()
print("The catch clause at _precompile_worker line ~762:")
print("    except (OutOfResources, PTXASError, IntelGPUError) as e:")
print()
print("CompilationError is NOT in this list.")
print("A single invalid config crashes ALL precompilation.")
```

**Reproducer showing the Triton broadcast failure (still live on tl.atomic_add):**

```python
"""
tl.atomic_add still uses broadcast_impl_shape (1-way) even in Triton 3.7.0.
This fails when ptr shape [1, RBLOCK] != mask shape [XBLOCK, 1] and XBLOCK > 1.
"""
import triton
import triton.language as tl
import torch

@triton.jit
def kernel_atomic_mismatch(out_ptr, val_ptr, xnumel, XBLOCK: tl.constexpr):
    R0_BLOCK: tl.constexpr = 16
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    r0_index = tl.arange(0, R0_BLOCK)[None, :]

    val = tl.load(val_ptr + r0_index, None)

    # ptr shape: [1, R0_BLOCK] (only depends on r0_index)
    # mask shape: [XBLOCK, 1] (depends on xindex)
    # broadcast_impl_shape([XBLOCK, 1], [1, R0_BLOCK]) fails when XBLOCK > 1
    tl.atomic_add(out_ptr + r0_index, val, xmask, sem='relaxed')


if __name__ == "__main__":
    print(f"Triton version: {triton.__version__}")
    out = torch.zeros(16, device='cuda')
    val = torch.ones(16, device='cuda')

    for xblock in [1, 8]:
        try:
            grid = ((32 + xblock - 1) // xblock,)
            out.zero_()
            kernel_atomic_mismatch[grid](out, val, 32, XBLOCK=xblock)
            torch.cuda.synchronize()
            print(f"XBLOCK={xblock}: SUCCESS")
        except Exception as e:
            err = str(e)
            if 'broadcast' in err.lower():
                print(f"XBLOCK={xblock}: BROADCAST ERROR (CompilationError)")
            else:
                print(f"XBLOCK={xblock}: {type(e).__name__}: {err[:200]}")
```

**Output:**
```
Triton version: 3.7.0
XBLOCK=1: SUCCESS
XBLOCK=8: BROADCAST ERROR (CompilationError)
```

### Expected behavior

`_precompile_worker` should catch `CompilationError` (from `triton.compiler.errors`) alongside the existing `OutOfResources`, `PTXASError`, and `IntelGPUError`, and skip that config rather than crashing. If ALL configs fail, it should raise `NoTritonConfigsError` with the last exception info, just like it does for `OutOfResources` today.

### Environment

- PyTorch: 2.14.0a0+git22b5803 (nightly)
- Triton: 3.7.0
- CUDA: 12.8
- Python: 3.12.12
- GPU: NVIDIA H200 (cc=9.0)
- OS: Linux (RHEL 9 / glibc 2.40)

### Related issues/PRs

- triton-lang/triton#7988 — Fixed `tl.load` broadcast (Triton v3.5.0)
- triton-lang/triton#8661 — Fixed `tl.store` broadcast (Triton v3.6.0)
- `tl.atomic_add` broadcast (see triton-lang/triton#10896)
- pytorch/pytorch#164700 — Related Inductor `broadcast_to` crash (fixed in codegen)
- pytorch/pytorch#172658 — Related combo kernel broadcast mismatch (fixed in codegen)
- pytorch/pytorch#141744 — Related DeviceProperties portability issue

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @ydwu4 @desertfire @yushangdi @iupaikov-amd @jansel @eellison @oulgen @shunting314

**Comment by CodersAcademy006** [comment]
Confirmed: in `_precompile_worker` (`torch/_inductor/runtime/triton_heuristics.py`, ~L756-764) the per-config compile loop only catches `(OutOfResources, PTXASError, IntelGPUError)` at L759, so a config whose generated Triton is structurally invalid raises `triton.compiler.errors.CompilationError`, which propagates out and aborts the whole precompile. During a normal `torch.compile` run this never surfaces because a working config is selected and the bad one is only compiled lazily; on cold AOT import every config is compiled up front, so the first invalid one crashes the load. Adding `CompilationError` to the caught tuple (it isn't currently imported in this file, so it needs the import too) would let it fall through to the existing "no valid configs" handling at L761, which already raises cleanly only when *all* configs fail. @XAheli

**Comment by zou3519** [comment]
cc @eellison  @aorenste
